### PR TITLE
Improve JUnitXmlTestRunResultProcessor's URL generation

### LINKED ISSUE #21: Improper characters being used to construct URL in JUnitXmlTestRunResultProcessor
This was logged by continuous integration:

```
WARNING: Invocation of TestRunResultProcessor org.igniterealtime.smack.inttest.util.JUnitXmlTestRunResultProcessor@6562cc23 failed.
java.lang.IllegalArgumentException: Illegal character in fragment at index 45: https://xmpp.org/extensions/xep-0045.html#7.1 & 7.2.2
	at java.base/java.net.URI.create(URI.java:883)
	at org.igniterealtime.smack.inttest.util.JUnitXmlTestRunResultProcessor.process(JUnitXmlTestRunResultProcessor.java:215)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.main(SmackIntegrationTestFramework.java:119)
Caused by: java.net.URISyntaxException: Illegal character in fragment at index 45: https://xmpp.org/extensions/xep-0045.html#7.1 & 7.2.2
	at java.base/java.net.URI$Parser.fail(URI.java:2913)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
	at java.base/java.net.URI$Parser.parse(URI.java:3128)
	at java.base/java.net.URI.<init>(URI.java:600)
	at java.base/java.net.URI.create(URI.java:881)
	... 2 more
```
It appears that some content that's used to generate a URL is not properly escaped.

### PR DESCRIPTION
**Pull request by guusdk**
The processor that generates JUnit-compatible XML reports attempts to generate an URL to the specifications that are under test.

Input used for these URLs should be properly escaped.

Links should be generated not only for XEP-specifications, but also for RFCs.

If a link can't be created, prevent the report generation from crashing. Instead, just generate a report without a link.

fixes #21

### CODE REVIEW

File: src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java [hunk 1]
```diff
@@ -239,6 +252,53 @@ public void process(SmackIntegrationTestFramework.TestRunResult testRunResult)
...
     }
 
+    public static URI generateLink(final String specification, final String specificationSection) {
+        if (specification == null || specification.isBlank()) {
+            return null;
+        }
+
+        String link;
+        if (specification.toUpperCase().startsWith("XEP")) {
+            String normalizedSpec = specification.substring(0, 3);
```
Comment (Fishbowler) [inline src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java:262]: You want to preserve the author's casing here?

File: src/test/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        final String specification = applySmackNormalization("xep-0030");
+        final String specificationSection = null;
+
+        // Execute system under test.
+        final URI result = JUnitXmlTestRunResultProcessor.generateLink(specification, specificationSection);
+
+        // Verify results.
+        assertNotNull(result);
+        assertEquals("https://xmpp.org/extensions/xep-0030.html", result.toString());
+    }
```
Comment (Fishbowler) [inline src/test/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessorTest.java:133]: Is this the same as testXepWithoutSection?

**Review by Fishbowler — COMMENTED** [review]
I've added a couple of comments.

I'm not sure that these tests prove the right thing. They prove the method works when its been through that normalisation method, but is that true in the normal execution? My tracing couldn't find anywhere it's used outside of the collection at the end.

File: src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java (hunk 1 above)
Comment (guusdk) [reply src/main/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessor.java:262]: Yes, I do. I'm expected smack-normalized input (although I don't know if that's lower or uppercase).

File: src/test/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessorTest.java (hunk 2 above)
Comment (guusdk) [reply src/test/java/org/igniterealtime/smack/inttest/util/JUnitXmlTestRunResultProcessorTest.java:133]: Hehe, possibly. It's me going through all permutations in my head.

Still, there might be value to keep this. If we ever end up modifying its duplicate for reasons© then it's good to have this explicit test that uses a dash, which is documented to be an integral part of the test by the method name. Execution-wise, the duplication costs next to nothing.

**Comment by guusdk** [comment]
> They prove the method works when its been through that normalisation method, but is that true in the normal execution?

You might be right. I assumed it was, but on a closer look, I believe that Smack only normalises when it applies the enabled/disabled-specification configuration.

I'm thinking that it would be kind of nice if Smack's output would actually contain normalized references, as that would potentially prevent a lot of follow-up issues. I'll see if I can add that to Smack.

**Review by Fishbowler — APPROVED** [review]

**Pull request merged**